bin/proving-service/src/proxy/status.rs

```diff
@@ -0,0 +1,128 @@
+    connection_timeout: Duration,
+    total_timeout: Duration,
+) -> Result<StatusApiClient<Channel>, ProvingServiceError> {
+    let channel = Channel::from_shared(format!("http://{}", address))
+        .map_err(|err| ProvingServiceError::InvalidURI(err, address.clone()))?
+        .connect_timeout(connection_timeout)
+        .timeout(total_timeout)
+        .connect()
+        .await
+        .map_err(|err| ProvingServiceError::ConnectionFailed(err, address))?;
+
+    Ok(StatusApiClient::new(channel))
+}
+
+// Status of a worker
+#[derive(Debug, Serialize)]
+pub struct WorkerStatus {
+    address: String,
+    version: String,
+    status: WorkerHealthStatus,
+}
+
+/// Status of the proxy
+#[derive(Debug, Serialize)]
+pub struct ProxyStatus {
+    version: String,
+    supported_proof_type: ProverType,
+    busy_workers: usize,
+    workers: Vec<WorkerStatus>,
+}
+
+/// Service that handles status requests
+pub struct StatusService {
+    load_balancer: Arc<LoadBalancerState>,
+}
+
+impl StatusService {
+    pub fn new(load_balancer: Arc<LoadBalancerState>) -> Self {
+        Self { load_balancer }
+    }
+
+    async fn handle_request(&self, _session: &mut ServerSession) -> Result<()> {
+        let workers = self.load_balancer.workers.read().await;
+        let worker_statuses: Vec<WorkerStatus> = workers
+            .iter()
+            .map(|w| WorkerStatus {
+                address: w.address(),
+                version: w.version().to_string(),
+                status: w.health_status().clone(),
+            })
+            .collect();
+
+        let status = ProxyStatus {
+            version: env!("CARGO_PKG_VERSION").to_string(),
+            supported_proof_type: self.load_balancer.supported_prover_type,
+            busy_workers: workers.iter().filter(|w| !w.is_available()).count(),
+            workers: worker_statuses,
+        };
+
+        let response = serde_json::to_string(&status).map_err(|e| {
+            Error::explain(
+                ErrorType::Custom("Failed to serialize status"),
+                format!("Failed to serialize status: {}", e),
+            )
+        })?;
+
+        let mut header = ResponseHeader::build(200, None)?;
+        header.insert_header("Content-Type", "application/json")?;
+        _session.write_response_header(Box::new(header)).await?;
+        _session.write_response_body(response.into(), true).await?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl HttpServerApp for StatusService {
+    async fn process_new_http(
+        self: &Arc<Self>,
+        mut session: ServerSession,
+        _shutdown: &ShutdownWatch,
+    ) -> Option<Stream> {
+        match session.read_request().await {
+            Ok(false) => return None,
+            Err(e) => {
+                error!("Failed to read request: {}", e);
+                return None;
+            },
+            Ok(true) => {},
+        }
+
+        if session.req_header().uri.path() != "/status" {
+            return None;
```

Comment:
Maybe not for now but should this return an error?

Reply:
I changed it to return a 404